File: code/client/munkilib/updatecheck/installationstate.py
# encoding: utf-8
#
# Copyright 2009-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
updatecheck.installationstate

Created by Greg Neagle on 2017-01-01.

Utilities for determining installation status for Munki items.
"""
# This code is largely still compatible with Python 2, so for now, turn off
# Python 3 style warnings
# pylint: disable=consider-using-f-string
# pylint: disable=redundant-u-string-prefix

from __future__ import absolute_import, print_function

import os

from . import catalogs
from . import compare

from .. import display
from .. import osutils
from .. import profiles
from .. import scriptutils
from .. import utils
from ..wrappers import unicode_or_str


def installed_state(item_pl):
    """Checks to see if the item described by item_pl (or a newer version) is
    currently installed

    All tests must pass to be considered installed.
    Returns 1 if it looks like this version is installed
    Returns 2 if it looks like a newer version is installed.
    Returns 0 otherwise.
    """
    foundnewer = False

    if item_pl.get('OnDemand'):
        # always install these items -- retcode 0 means install is needed
        display.display_debug1('This is an OnDemand item. Must install.')
        return 0

    if item_pl.get('installcheck_script'):
        retcode = scriptutils.run_embedded_script(
            'installcheck_script', item_pl, suppress_error=True)
        display.display_debug1('installcheck_script returned %s', retcode)
        # retcode 0 means install is needed
        if retcode == 0:
            return 0
        # non-zero could be an error or successfully indicating
        # that an install is not needed. We hope it's the latter.
        # return 1 so we're marked as not needing to be installed
        return 1

    # this was deprecated a very long time ago. removing 02 Jan 2017
    #if item_pl.get('softwareupdatename'):
    #    available_apple_updates = appleupdates.softwareUpdateList()
    #    display.display_debug2(
    #        'Available Apple updates:\n%s', available_apple_updates)
    #    if item_pl['softwareupdatename'] in available_apple_updates:
    #        display.display_debug1(
    #            '%s is in available Apple Software Updates',
    #            item_pl['softwareupdatename'])
    #        # return 0 so we're marked as needing to be installed
    #        return 0
    #    else:
    #        display.display_debug1(
    #            '%s is not in available Apple Software Updates',
    #            item_pl['softwareupdatename'])
    #        # return 1 so we're marked as not needing to be installed
    #        return 1

    if item_pl.get('installer_type') == 'startosinstall':
        current_os_vers = osutils.getOsVersion()
        item_os_vers = item_pl.get('version')
        if int(item_os_vers.split('.')[0]) > 10:
            # if we're running Big Sur+, we just want the major (11)
            item_os_vers = item_os_vers.split('.')[0]
        else:
            # need just major.minor part of the version -- 10.12 and not 10.12.4
            item_os_vers = '.'.join(item_os_vers.split('.')[0:2])
        comparison = compare.compare_versions(current_os_vers, item_os_vers)
        if comparison == compare.VERSION_IS_LOWER:
            return 0
        if comparison == compare.VERSION_IS_HIGHER:
            return 2
        # version is the same
        return 1

    if item_pl.get('installer_type') == 'stage_os_installer':
        # we return 2 if the installed macOS is the same version or higher than
        # the version of this item
        # we return 1 if the OS installer has already been staged
        # otherwise return 0
        current_os_vers = osutils.getOsVersion()
        item_os_vers = item_pl.get('version')
        if int(item_os_vers.split('.')[0]) > 10:
            # if we're running Big Sur+, we just want the major (11)
            item_os_vers = item_os_vers.split('.')[0]
        else:
            # need just major.minor part of the version -- 10.12 and not 10.12.4
            item_os_vers = '.'.join(item_os_vers.split('.')[0:2])
        comparison = compare.compare_versions(current_os_vers, item_os_vers)
        if comparison in (compare.VERSION_IS_THE_SAME, compare.VERSION_IS_HIGHER):
            return 2
        # installed OS version is lower; check to see if we've staged the installer
        for item in item_pl.get("installs", []):
            try:
                comparison = compare.compare_item_version(item)
                if comparison == compare.VERSION_IS_THE_SAME:
                    return 1
                #else
                return 0
            except utils.Error as err:
                # some problem with the installs data
                display.display_error(unicode_or_str(err))
                # return 1 so we're marked as not needing to be installed
                return 1

    if item_pl.get('installer_type') == 'profile':
        identifier = item_pl.get('PayloadIdentifier')
        hash_value = item_pl.get('installer_item_hash')
        if profiles.profile_needs_to_be_installed(identifier, hash_value):
            return 0
        # does not need to be installed
        return 1

    # does 'installs' exist and is it non-empty?
    if item_pl.get('installs', None):
        installitems = item_pl['installs']
        for item in installitems:
            try:
                comparison = compare.compare_item_version(item)
                if comparison in (-1, 0):
                    return 0
                if comparison == 2:
                    # this item is newer
                    foundnewer = True
            except utils.Error as err:
                # some problem with the installs data
                display.display_error(unicode_or_str(err))
                # return 1 so we're marked as not needing to be installed
                return 1

    # if there is no 'installs' key, then we'll use receipt info
    # to determine install status.
    elif 'receipts' in item_pl:
        receipts = item_pl['receipts']
        for item in receipts:
            try:
                comparison = compare.compare_receipt_version(item)
                if comparison in (-1, 0):
                    # not there or older
                    return 0
                if comparison == 2:
                    foundnewer = True
            except utils.Error as err:
                # some problem with the receipts data
                display.display_error(unicode_or_str(err))
                # return 1 so we're marked as not needing to be installed
                return 1

    # if we got this far, we passed all the tests, so the item
    # must be installed (or we don't have enough info...)
    if foundnewer:
        return 2
    # not newer
    return 1


def some_version_installed(item_pl):
    """Checks to see if some version of an item is installed.

    Args:
      item_pl: item plist for the item to check for version of.

    Returns a boolean.
    """
    if item_pl.get('OnDemand'):
        # These should never be counted as installed
        display.display_debug1('This is an OnDemand item.')
        return False

    if item_pl.get('installcheck_script'):
        retcode = scriptutils.run_embedded_script(
            'installcheck_script', item_pl, suppress_error=True)
        display.display_debug1(
            'installcheck_script returned %s', retcode)
        # retcode 0 means install is needed
        # (ie, item is not installed)
        if retcode == 0:
            return False
        # non-zero could be an error or successfully indicating
        # that an install is not needed. We hope it's the latter.
        return True

    if item_pl.get('installer_type') in ['startosinstall', 'stage_os_installer']:
        # Some version of macOS is always installed!
        return True

    if item_pl.get('installer_type') == 'profile':
        identifier = item_pl.get('PayloadIdentifier')
        return profiles.profile_is_installed(identifier)

    # does 'installs' exist and is it non-empty?
    if item_pl.get('installs'):
        installitems = item_pl['installs']
        # check each item for existence
        for item in installitems:
            try:
                if compare.compare_item_version(item) == 0:
                    # not there
                    return False
            except utils.Error as err:
                # some problem with the installs data
                display.display_error(unicode_or_str(err))
                return False

    # if there is no 'installs' key, then we'll use receipt info
    # to determine install status.
    elif 'receipts' in item_pl:
        receipts = item_pl['receipts']
        for item in receipts:
            try:
                if compare.compare_receipt_version(item) == 0:
                    # not there
                    return False
            except utils.Error as err:
                # some problem with the installs data
                display.display_error(unicode_or_str(err))
                return False

    # if we got this far, we passed all the tests, so the item
    # must be installed (or we don't have enough info...)
    return True


def evidence_this_is_installed(item_pl):
    """Checks to see if there is evidence that the item described by item_pl
    (any version) is currently installed.

    If any tests pass, the item might be installed.
    This is used when determining if we can remove the item, thus
    the attention given to the uninstall method.

    Returns a boolean.
    """
    if item_pl.get('OnDemand'):
        # These should never be counted as installed
        display.display_debug1('This is an OnDemand item.')
        return False

    if item_pl.get('uninstallcheck_script'):
        retcode = scriptutils.run_embedded_script(
            'uninstallcheck_script', item_pl, suppress_error=True)
        display.display_debug1(
            'uninstallcheck_script returned %s', retcode)
        # retcode 0 means uninstall is needed
        # (ie, item is installed)
        if retcode == 0:
            return True
        # non-zero could be an error or successfully indicating
        # that an uninstall is not needed
        return False

    if item_pl.get('installcheck_script'):
        retcode = scriptutils.run_embedded_script(
            'installcheck_script', item_pl, suppress_error=True)
        display.display_debug1(
            'installcheck_script returned %s', retcode)
        # retcode 0 means install is needed
        # (ie, item is not installed)
        if retcode == 0:
            return False
        # non-zero could be an error or successfully indicating
        # that an install is not needed
        return True

    if item_pl.get('installer_type') == 'startosinstall':
        # Some version of macOS is always installed!
        return True

    if item_pl.get('installer_type') == 'profile':
        identifier = item_pl.get('PayloadIdentifier')
        return profiles.profile_is_installed(identifier)

    foundallinstallitems = False
    if ('installs' in item_pl and
            item_pl.get('uninstall_method') != 'removepackages'):
        display.display_debug2("Checking 'installs' items...")
        installitems = item_pl['installs']
        if installitems:
            foundallinstallitems = True
            for item in installitems:
                if 'path' in item:
                    # we can only check by path; if the item has been moved
                    # we're not clever enough to find it, and our removal
                    # methods are currently even less clever
                    if not os.path.exists(item['path']):
                        # this item isn't on disk
                        display.display_debug2(
                            '%s not found on disk.', item['path'])
                        foundallinstallitems = False
        if (foundallinstallitems and
                item_pl.get('uninstall_method') != 'removepackages'):
            return True
    if item_pl.get('receipts'):
        display.display_debug2("Checking receipts...")
        pkgdata = catalogs.analyze_installed_pkgs()
        if item_pl['name'] in pkgdata['installed_names']:
            return True
        #else:
        display.display_debug2("Installed receipts don't match.")

    # if we got this far, we failed all the tests, so the item
    # must not be installed (or we don't have the right info...)
    return False


if __name__ == '__main__':
    print('This is a library of support tools for the Munki Suite.')

File: code/cli/munki/shared/updatecheck/installationstate.swift
//
//  installationstate.swift
//  munki
//
//  Created by Greg Neagle on 8/19/24.
//  Copyright 2024-2026 The Munki Project. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//       https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

import Foundation

private let display = DisplayAndLog.main

enum InstallationState: Int {
    case thisVersionNotInstalled = 0
    case thisVersionInstalled = 1
    case newerVersionInstalled = 2
}

/// Checks to see if the item described by pkginfo (or a newer version) is
/// currently installed
///
/// All tests must pass to be considered installed.
/// Returns InstallationState
func installedState(_ pkginfo: PlistDict) async -> InstallationState {
    var foundNewer = false
    if pkginfo["OnDemand"] as? Bool ?? false {
        // we always need to install these items
        display.debug1("This is an OnDemand item. Must install.")
        return .thisVersionNotInstalled
    }
    if pkginfo["installcheck_script"] is String {
        let retcode = await runEmbeddedScript(
            name: "installcheck_script",
            pkginfo: pkginfo,
            suppressError: true
        )
        display.debug1("installcheck_script returned \(retcode)")
        // retcode 0 means install IS needed
        if retcode == 0 {
            return .thisVersionNotInstalled
        }
        // non-zero could be an error or successfully indicating
        // that an install is not needed. We hope it's the latter.
        // return .thisVersionInstalled so we're marked as not needing to be installed
        return .thisVersionInstalled
    }
    if pkginfo["version_script"] is String {
        // if a version_script is defined, use that to determine installedState
        let compareResult = await compareUsingVersionScript(pkginfo)
        if compareResult == .notPresent || compareResult == .older {
            return .thisVersionNotInstalled
        }
        if compareResult == .newer {
            return .newerVersionInstalled
        }
        if compareResult == .same {
            return .thisVersionInstalled
        }
    }
    let installerType = pkginfo["installer_type"] as? String ?? ""
    if installerType == "startosinstall",
       var installerItemVersion = pkginfo["version"] as? String
    {
        let currentOSVersion = getOSVersion() // just gets major.minor
        let installerVersionParts = installerItemVersion.components(separatedBy: ".")
        if (Int(installerVersionParts[0]) ?? 0) > 10 {
            // if we're running Big Sur+, we just want the major (11, 12, etc)
            installerItemVersion = installerVersionParts[0]
        } else {
            // need just major.minor part of the version -- 10.12 and not 10.12.4
            installerItemVersion = installerVersionParts[0] + "." + installerVersionParts[1]
        }
        let compareResult = compareVersions(currentOSVersion, installerItemVersion)
        if compareResult == .older {
            return .thisVersionNotInstalled
        }
        if compareResult == .newer {
            return .newerVersionInstalled
        }
        return .thisVersionInstalled
    }
    if installerType == "stage_os_installer",
       var installerItemVersion = pkginfo["version"] as? String
    {
        // we return .newerVersionInstalled if the installed macOS is the same version
        // or higher than the version of this item
        // we return .thisVersionInstalled if the OS installer has already been staged
        // otherwise return .thisVersionNotInstalled
        let currentOSVersion = getOSVersion() // just gets major.minor
        let installerVersionParts = installerItemVersion.components(separatedBy: ".")
        if (Int(installerVersionParts[0]) ?? 0) > 10 {
            // if we're running Big Sur+, we just want the major (11, 12, etc)
            installerItemVersion = installerVersionParts[0]
        } else {
            // need just major.minor part of the version -- 10.12 and not 10.12.4
            installerItemVersion = installerVersionParts[0] + "." + installerVersionParts[1]
        }
        let compareResult = compareVersions(currentOSVersion, installerItemVersion)
        if compareResult == .same || compareResult == .newer {
            return .newerVersionInstalled
        }
        // installed OS version is lower; check to see if we've staged the os installer
        for item in pkginfo["installs"] as? [PlistDict] ?? [] {
            do {
                let compareResult = try compareItem(item)
                if compareResult != .same {
                    return .thisVersionNotInstalled
                }
            } catch {
                display.error(error.localizedDescription)
                // return .thisVersionInstalled so we don't attempt an install
                return .thisVersionInstalled
            }
        }
        // all items are present and same version
        return .thisVersionInstalled
    }
    // do we have installs items?
    if let installItems = pkginfo["installs"] as? [PlistDict],
       !installItems.isEmpty
    {
        for item in installItems {
            do {
                let compareResult = try compareItem(item)
                if compareResult == .older || compareResult == .notPresent {
                    return .thisVersionNotInstalled
                }
                if compareResult == .newer {
                    foundNewer = true
                }
            } catch {
                display.error(error.localizedDescription)
                // return .thisVersionInstalled so we don't attempt an install
                return .thisVersionInstalled
            }
        }
    } else if let receipts = pkginfo["receipts"] as? [PlistDict] {
        // if there are no 'installs' items, then we'll use receipt info
        // to determine install status.
        for item in receipts {
            do {
                let compareResult = try await compareReceipt(item)
                if compareResult == .older || compareResult == .notPresent {
                    return .thisVersionNotInstalled
                }
                if compareResult == .newer {
                    foundNewer = true
                }

            } catch {
                display.error(error.localizedDescription)
                // return .thisVersionInstalled so we don't attempt an install
                return .thisVersionInstalled
            }
        }
    }
    // if we got this far, we passed all the tests, so the item
    // must be installed (or we don't have enough info...)
    if foundNewer {
        return .newerVersionInstalled
    }
    return .thisVersionInstalled
}

/// Checks to see if some version of a pkgitem is installed.
func someVersionInstalled(_ pkginfo: PlistDict) async -> Bool {
    if pkginfo["OnDemand"] as? Bool ?? false {
        // These should never be counted as installed
        display.debug1("This is an OnDemand item.")
        return false
    }
    if pkginfo["installcheck_script"] is String {
        // installcheck_script can really only tell us that an item needs
        // to be installed, or it doesn't.
        // it can't tell us that an older version of the item is installed
        let retcode = await runEmbeddedScript(
            name: "installcheck_script",
            pkginfo: pkginfo,
            suppressError: true
        )
        display.debug1("installcheck_script returned \(retcode)")
        // retcode 0 means install is needed
        // (ie, item is not installed)
        // non-zero could be an error or successfully indicating
        // that an install is not needed. We hope it's the latter.
        return retcode != 0
    }
    if pkginfo["version_script"] is String {
        // if there's a version_script, let's use that to determine
        // if some version installed
        let comparisonResult = await compareUsingVersionScript(pkginfo)
        if comparisonResult == .notPresent {
            return false
        }
        return true
    }
    if let installerType = pkginfo["installer_type"] as? String,
       installerType == "startosinstall" || installerType == "stage_os_installer"
    {
        // Some version of macOS is always installed!
        return true
    }
    // do we have installs items?
    if let installItems = pkginfo["installs"] as? [PlistDict],
       !installItems.isEmpty
    {
        for item in installItems {
            do {
                let compareResult = try compareItem(item)
                if compareResult == .notPresent {
                    return false
                }
            } catch {
                display.error(error.localizedDescription)
                return false
            }
        }
    } else if let receipts = pkginfo["receipts"] as? [PlistDict] {
        for item in receipts {
            do {
                let compareResult = try await compareReceipt(item)
                if compareResult == .notPresent {
                    return false
                }

            } catch {
                display.error(error.localizedDescription)
                // return .thisVersionInstalled so we don't attempt an install
                return false
            }
        }
    }
    // if we got this far, we passed all the tests, so the item
    // must be installed (or we don't have enough info...)
    return true
}

/// Checks to see if there is any evidence that the item described
/// by pkginfo (any version) is currently installed.
/// If any tests pass, the item might be installed.
/// This is used when determining if we can remove the item, thus
/// the attention given to the uninstall method.
func evidenceThisIsInstalled(_ pkginfo: PlistDict) async -> Bool {
    if pkginfo["OnDemand"] as? Bool ?? false {
        // These should never be counted as installed
        display.debug1("This is an OnDemand item.")
        return false
    }
    if pkginfo["uninstallcheck_script"] is String {
        // installcheck_script can really only tell us that an item needs
        // to be installed, or it doesn't.
        // it can't tell us that an older version of the item is installed
        let retcode = await runEmbeddedScript(
            name: "uninstallcheck_script",
            pkginfo: pkginfo,
            suppressError: true
        )
        display.debug1("uninstallcheck_script returned \(retcode)")
        // retcode 0 means uninstall is needed
        // (ie, item is installed)
        // non-zero could be an error or successfully indicating
        // that an uninstall is not needed.
        return retcode == 0
    }
    if pkginfo["installcheck_script"] is String {
        // installcheck_script can really only tell us that an item needs
        // to be installed, or it doesn't.
        // it can't tell us that an older version of the item is installed
        let retcode = await runEmbeddedScript(
            name: "installcheck_script",
            pkginfo: pkginfo,
            suppressError: true
        )
        display.debug1("installcheck_script returned \(retcode)")
        // retcode 0 means install is needed
        // (ie, item is not installed)
        // non-zero could be an error or successfully indicating
        // that an install is not needed. We hope it's the latter.
        return retcode != 0
    }
    if pkginfo["version_script"] is String {
        // if a comparison using a version_script returns anything
        // other than .notPresent that's evidence the item is installed
        let comparisonResult = await compareUsingVersionScript(pkginfo)
        if comparisonResult != .notPresent {
            return true
        }
    }
    if let installerType = pkginfo["installer_type"] as? String,
       installerType == "startosinstall" || installerType == "stage_os_installer"
    {
        // Some version of macOS is always installed!
        return true
    }
    var foundAllInstallItems = false
    if let installItems = pkginfo["installs"] as? [PlistDict],
       !installItems.isEmpty,
       (pkginfo["uninstall_method"] as? String ?? "") != "removepackages"
    {
        display.debug2("Checking 'installs' items...")
        foundAllInstallItems = true
        for item in installItems {
            if let path = item["path"] as? String, !pathExists(path) {
                // this item isn't present
                display.debug2("\(path) not found on disk.")
                foundAllInstallItems = false
            }
        }
        if foundAllInstallItems {
            display.debug2("Found all installs items")
            return true
        }
    }
    if let itemName = pkginfo["name"] as? String,
       let receipts = pkginfo["receipts"] as? [PlistDict],
       !receipts.isEmpty
    {
        display.debug2("Checking receipts...")
        let pkgdata = await analyzeInstalledPkgs()

        if let installedNames = pkgdata["installed_names"] as? [String],
           installedNames.contains(itemName)
        {
            display.debug2("Found matching receipts")
            return true
        }
        display.debug2("Installed receipts don't match for \(itemName)")
    }
    // if we got this far, we failed all the tests, so the item
    // must not be installed (or we don't have the right info...)
    return false
}
